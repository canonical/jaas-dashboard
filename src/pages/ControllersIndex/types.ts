export enum Label {
  DEFAULT = "Default",
  NAME = "Name",
}

export enum TestId {
  COMPONENT = "ControllersIndex",
}
