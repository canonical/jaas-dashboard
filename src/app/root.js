import immerProduce from "immer";
import cloneDeep from "clone-deep";

import { actionsList } from "./action-types";

function rootReducer(state = {}, action) {
  return immerProduce(state, (draftState) => {
    switch (action.type) {
      case actionsList.updateControllerConnection:
        const connections = cloneDeep(state.controllerConnections || {});
        connections[action.payload.wsControllerURL] = action.payload.info;
        draftState.controllerConnections = connections;
        break;
      case actionsList.storeConfig:
        draftState.config = action.payload;
        break;
      case actionsList.storeLoginError:
        draftState.loginError = action.payload;
        break;
      case actionsList.storeUserPass:
        const credentials = cloneDeep(state.credentials || {});
        credentials[action.payload.wsControllerURL] = action.payload.credential;
        draftState.credentials = credentials;
        break;
      case actionsList.storeVersion:
        draftState.appVersion = action.payload;
        break;
      case actionsList.storeVisitURL:
        draftState.visitURL = action.payload;
        break;
      case actionsList.logOut:
        delete draftState.controllerConnections;
        delete draftState.visitURL;
        break;
      case actionsList.updatePingerIntervalId:
        const intervals = cloneDeep(state.pingerIntervalIds || {});
        intervals[action.payload.wsControllerURL] = action.payload.intervalId;
        draftState.pingerIntervalIds = intervals;
        break;
      default:
        // no default value, fall through.
        break;
    }
  });
}

export default rootReducer;
