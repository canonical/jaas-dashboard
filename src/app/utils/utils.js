import { parseISO, formatDistanceToNow } from "date-fns";
import { useState } from "react";
import defaultCharmIcon from "static/images/icons/default-charm-icon.svg";

export function generateEntityIdentifier(charmId, name, subordinate) {
  if (!charmId) {
    return null;
  }

  return (
    <div className="entity-name u-truncate" title={name}>
      {subordinate && <span className="subordinate"></span>}
      {charmId && generateIconImg(name, charmId)}
      {name}
    </div>
  );
}

export const generateStatusElement = (
  status = "unknown",
  count,
  useIcon = true,
  actionsLogs = false
) => {
  let statusClass = status ? `is-${status.toLowerCase()}` : "";
  let countValue = "";
  if (count !== undefined) {
    countValue = ` (${count})`;
  }
  const className = useIcon ? "status-icon " + statusClass : "";
  // ActionLogs uses a spinner icon if 'running'
  if (actionsLogs && statusClass === "is-running") {
    return (
      <span className="status-icon is-spinner">
        <i className="p-icon--spinner u-animation--spin"></i>
        {status}
      </span>
    );
  }

  return (
    <span className={className}>
      {status}
      {countValue}
    </span>
  );
};

export const generateSpanClass = (className, value) => {
  return <span className={className}>{value}</span>;
};

// Highest status to the right
const statusOrder = ["running", "alert", "blocked"];

const setHighestStatus = (entityStatus, highestStatus) => {
  if (statusOrder.indexOf(entityStatus) > statusOrder.indexOf(highestStatus)) {
    return entityStatus;
  }
  return highestStatus;
};

// If it's the highest status then we don't need to continue looping
// applications or units.
const checkHighestStatus = (highestStatus) => {
  return highestStatus === statusOrder[statusOrder.length - 1];
};

export const getModelStatusGroupData = (model) => {
  let highestStatus = statusOrder[0]; // Set the highest status to the lowest.
  let messages = [];
  const applications = model.applications || {};
  Object.keys(applications).forEach((appName) => {
    const app = applications[appName];
    const { status: appStatus } = getApplicationStatusGroup(app);
    highestStatus = setHighestStatus(appStatus, highestStatus);
    if (checkHighestStatus(highestStatus)) {
      // If it's the highest status then we want to store the message.
      messages.push(app.status.info);
      return;
    }
    const units = app.units || {}; // subordinates do not have units.
    Object.keys(units).forEach((unitId) => {
      const unit = units[unitId];
      const { status: unitStatus } = getUnitStatusGroup(unit);
      highestStatus = setHighestStatus(unitStatus, highestStatus);
      if (checkHighestStatus(highestStatus)) {
        // If it's the highest status then we want to store the message.
        messages.push(unit["agent-status"].info);
        return;
      }
    });
  });
  return {
    highestStatus,
    messages,
  };
};

/**
  Returns the status for the application.
  @param {Object} application The application to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the application and any relevent messaging.
*/
export const getApplicationStatusGroup = (application) => {
  // Possible "blocked" or error states in application statuses.
  const blocked = ["blocked"];
  // Possible "alert" states in application statuses.
  const alert = ["unknown"];
  const status = application.status.status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns the status level for the machine.
  @param {Object} machine The machine to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the machine and any relevent messaging.
*/
export const getMachineStatusGroup = (machine) => {
  // Possible "blocked" or error states in machine statuses.
  const blocked = ["down"];
  // Possible "alert" states in machine statuses.
  const alert = ["pending"];
  const status = machine["agent-status"].status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns the status level for the unit.
  @param {Object} unit The unit to check the status of in the
    format stored in the redux store.
  @returns {Object} The status of the unit and any relevent messaging.
*/
export const getUnitStatusGroup = (unit) => {
  // Possible "blocked" or error states in the unit statuses.
  const blocked = ["lost"];
  // Possible "alert" states in the unit statuses.
  const alert = ["allocating"];
  const status = unit["agent-status"].status;
  const response = {
    status: "running",
    message: null,
  };
  if (blocked.includes(status)) {
    response.status = "blocked";
  }
  if (alert.includes(status)) {
    response.status = "alert";
  }
  return response;
};

/**
  Returns owner string from ownerTag
  @param {string} ownerTag The ownerTag identifier returns from the API
  @returns {string} The simplified owner string
*/
export const extractOwnerName = (tag) => {
  return tag.split("@")[0].replace("user-", "");
};

/**
  Pluralizes the supplied word based on the provided dataset.
  @param {Number} value The integer to be checked
  @param {string} string The item name to be pluralized
  @returns {string} The item pluralized if required
*/
export const pluralize = (value, string) => {
  const special = {
    active: "active",
    allocating: "allocating",
    down: "down",
    joined: "joined",
    lost: "lost",
    running: "running",
    started: "started",
    unknown: "unknown",
    waiting: "waiting",
  };
  if (value === 1) {
    return string;
  } else if (special[string]) {
    return special[string];
  }
  return `${string}s`;
};

/**
  Returns cloud string from cloudTag
  @param {string} cloudTag The cloudTag identifier returns from the API
  @returns {string} The simplified cloud string
*/
export const extractCloudName = (tag = "") => {
  return tag.replace("cloud-", "");
};

/**
  Returns credential string from Tag
  @param {string} cloudTag The cloudTag identifier returns from the API
  @returns {string} The simplified cloud string
*/
export const extractCredentialName = (tag) => {
  // @ is not there in local boostraps
  // cloudcred-localhost_admin_localhost
  if (!tag) return "-";
  let cred = tag.split("cloudcred-")[1];
  if (cred.indexOf("@") > -1) {
    return cred.split("@")[1].split("_")[1];
  }
  return cred.split("_")[1];
};

/**
  Returns the version of the supplied charm string.
  @param {String} charmName The full path of the charm e.g. cs:foo/bar-123
*/
export const extractRevisionNumber = (charmName = "") =>
  charmName.split("-").pop();

/**
  Returns a link to the charm icon for the provided charm name.
  @param {String} charmId The fully qualified charm name.
  @returns {String} The link to the charm icon.
*/
export const generateIconPath = (charmId) => {
  if (charmId.indexOf("local:") === 0) {
    return defaultCharmIcon;
  }
  if (charmId.indexOf("cs:") === 0) {
    // Strip unnessesary prefixes
    // before: cs:~containers/lxd-container-47
    // after:  containers/lxd-container-47
    charmId = charmId.replace("cs:~", "");
    charmId = charmId.replace("cs:", "");

    // Remove release from the charmId
    // before: containers/trusty/lxd-container-47
    // after:  containers/lxd-container-47
    charmId = charmId.replace("precise/", "");
    charmId = charmId.replace("trusty/", "");
    charmId = charmId.replace("xenial/", "");
    charmId = charmId.replace("bionic/", "");
    charmId = charmId.replace("focal/", "");

    // Combine owner and charm name
    // before: containers/lxd-container-47
    // after:  containers-lxd-container-47
    charmId = charmId.replace("/", "-");

    // Strip the revision number from the end
    // before: containers-lxd-container-47
    // after:  containers-lxd-container
    charmId = charmId.substr(0, charmId.lastIndexOf("-", charmId.length));

    return `https://charmhub.io/${charmId}/icon`;
  }
  if (charmId.indexOf("ch:") === 0) {
    // Regex explanation:
    // "ch:amd64/xenial/content-cache-425".match(/\/(.+)\/(.+)-\d+/)
    // Array(3) [ "/xenial/content-cache-425", "xenial", "content-cache" ]
    const charmName = charmId.match(/\/(.+)\/(.+)-\d+/)[2];
    return `https://charmhub.io/${charmName}/icon`;
  }
  return "";
};

/**
  @returns {Int || 0} Returns the current viewport width
*/
export const getViewportWidth = () => {
  const de = document.documentElement;
  return Math.max(de.clientWidth, window.innerWidth || 0);
};

/**
 * @param {function} fn Function to debounce
 * @param {Number} wait Time in milliseconds to wait
 */
export const debounce = (fn, wait) => {
  let t;
  return function () {
    clearTimeout(t);
    t = setTimeout(() => fn.apply(this, arguments), wait);
  };
};

/**
  Returns a bool on whether a user can admin controllers
  @param {Obj} conn The user connection.
  @returns {Boolean} If they are an admin or not.
*/
export const userIsControllerAdmin = (conn) => {
  const controllerAccess = conn?.info?.user?.controllerAccess;
  return ["superuser", "admin"].includes(controllerAccess);
};

export const isSet = (val) => val || val !== undefined;

export const extractRelationEndpoints = (relation) => {
  const endpoints = {};
  relation.endpoints.forEach((endpoint) => {
    const role = endpoint.relation.role;
    endpoints[role] =
      endpoint["application-name"] + ":" + endpoint.relation.name;
    endpoints[`${role}ApplicationName`] = endpoint["application-name"];
  });
  return endpoints;
};

const ImgWithFallback = ({
  src,
  fallback = defaultCharmIcon,
  alt,
  ...props
}) => {
  const [imgSrc, setImgSrc] = useState(src);
  const onError = () => setImgSrc(fallback);

  return (
    <img
      alt={alt}
      src={imgSrc ? imgSrc : fallback}
      onError={onError}
      {...props}
    />
  );
};

export const generateIconImg = (name, charmId) => {
  let iconSrc = defaultCharmIcon;
  if (charmId.indexOf("local:") !== 0) {
    iconSrc = generateIconPath(charmId);
  }
  return (
    <ImgWithFallback
      alt={name + " icon"}
      key={`${name}-${charmId}`}
      title={name}
      width="24"
      height="24"
      className="entity-icon"
      src={iconSrc}
    />
  );
};

export const generateRelationIconImage = (applicationName, applications) => {
  const application = applications[applicationName];
  if (!application || !applicationName) {
    return;
  }
  return generateIconImg(applicationName, application["charm-url"]);
};

export const formatFriendlyDateToNow = (date) => {
  const formattedDate = formatDistanceToNow(parseISO(date));
  return formattedDate.concat(" ago");
};

export const canAdministerModelAccess = (userName, modelUsers) => {
  let hasPermission = false;
  const sharingAccess = ["admin", "write", "owner"];
  modelUsers &&
    modelUsers.forEach((userObj) => {
      if (userObj.user === userName && sharingAccess.includes(userObj.access)) {
        hasPermission = true;
      }
    });
  return hasPermission;
};
