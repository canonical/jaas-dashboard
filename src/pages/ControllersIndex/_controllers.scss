@import "vanilla-framework/scss/vanilla";

.controllers {
  td,
  th {
    &:first-child {
      min-width: 25%;
      white-space: nowrap;
    }
  }

  &--header {
    border-bottom: 1px solid $colors--light-theme--border-default;
    color: $color-mid-dark;
    display: inline-grid;
    font-size: 0.7656rem;
    font-weight: 400;
    grid-template-columns: repeat(2, 50%);
    margin-bottom: 1rem;
    padding: 0.25rem 0 0.5rem 1rem;
    text-transform: uppercase;
    width: 100%;

    .controllers__heading {
      color: $color-dark;
      padding-top: 0.8rem;
      position: relative;
      top: 0.5rem;
    }

    .controllers--register {
      text-align: right;
    }

    button {
      margin: 0;
      max-width: 15rem;
    }
  }

  &__status {
    // Display block so that the status gets truncated if it doesn't fit in the
    // cell.
    display: block !important;
  }

  .controllers--registered-tooltip {
    @include vf-icon-help($color-mid-light);

    margin-left: 0.4rem;
  }
}
