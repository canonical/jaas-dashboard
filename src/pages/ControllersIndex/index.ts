export { default } from "./ControllersIndex";
export {
  Label as ControllersIndexLabel,
  TestId as ControllersIndexTestId,
} from "./types";
