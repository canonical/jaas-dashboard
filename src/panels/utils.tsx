import { useCallback } from "react";

import {
  type SetParams,
  type QueryParams,
  useQueryParams,
} from "hooks/useQueryParams";

type data = {
  th: string;
  td: string;
};

export const generatePanelTableRows = (data: data[]) => {
  return data.map(({ th, td }) => {
    return (
      <tr className="panel__tr" key={th + td}>
        <th className="panel__th">{th}</th>
        <td className="panel__td">{td}</td>
      </tr>
    );
  });
};

// TODO: Add tests for usePanelQueryParams.
export const usePanelQueryParams = <P extends QueryParams>(
  defaultQueryParams: P
): [P, SetParams<P>, () => void] => {
  const [queryParams, setQueryParams] = useQueryParams<P>(defaultQueryParams);

  const handleRemovePanelQueryParams = useCallback(() => {
    const paramsToClear: Partial<P> = {};
    Object.keys(defaultQueryParams).forEach((param: keyof P) => {
      paramsToClear[param] = undefined;
    });
    setQueryParams(paramsToClear);
  }, [defaultQueryParams, setQueryParams]);

  return [queryParams, setQueryParams, handleRemovePanelQueryParams];
};
