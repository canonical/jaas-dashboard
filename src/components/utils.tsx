import classNames from "classnames";
import { parseISO, formatDistanceToNow } from "date-fns";
import { ImgHTMLAttributes, useState } from "react";
import defaultCharmIcon from "static/images/icons/default-charm-icon.svg";
import { generateIconPath } from "store/juju/utils/models";
import { ModelData } from "../juju/types";

export function generateEntityIdentifier(
  charmId: string,
  name: string,
  subordinate: boolean
) {
  if (!charmId) {
    return null;
  }

  return (
    <div className="entity-name u-truncate" title={name}>
      {subordinate && <span className="subordinate"></span>}
      {charmId && generateIconImg(name, charmId)}
      {name}
    </div>
  );
}

export const generateStatusElement = (
  status = "unknown",
  count?: number | null,
  useIcon = true,
  actionsLogs = false,
  className = null
) => {
  let statusClass = status ? `is-${status.toLowerCase()}` : "";
  let countValue = "";
  if (count || count === 0) {
    countValue = ` (${count})`;
  }
  const iconClasses = useIcon ? "status-icon " + statusClass : "";
  // ActionLogs uses a spinner icon if 'running'
  if (actionsLogs && statusClass === "is-running") {
    return (
      <span className="status-icon is-spinner">
        <i className="p-icon--spinner u-animation--spin"></i>
        {status}
      </span>
    );
  }

  return (
    <span className={classNames(iconClasses, className)}>
      {status}
      {countValue}
    </span>
  );
};

export const isSet = (val: unknown) => val || val !== undefined;

/**
  @returns {Int || 0} Returns the current viewport width
*/
export const getViewportWidth = () => {
  const de = document.documentElement;
  return Math.max(de.clientWidth, window.innerWidth || 0);
};

const ImgWithFallback = ({
  src,
  fallback = defaultCharmIcon,
  alt,
  ...props
}: {
  src: string;
  fallback?: string;
  alt: string;
} & Partial<ImgHTMLAttributes<HTMLImageElement>>) => {
  const [imgSrc, setImgSrc] = useState(src);
  const onError = () => setImgSrc(fallback);

  return (
    <img
      alt={alt}
      src={imgSrc ? imgSrc : fallback}
      onError={onError}
      {...props}
    />
  );
};

export const generateIconImg = (name: string, charmId: string) => {
  let iconSrc = defaultCharmIcon;
  if (charmId.indexOf("local:") !== 0) {
    iconSrc = generateIconPath(charmId);
  }
  return (
    <ImgWithFallback
      alt={name + " icon"}
      key={`${name}-${charmId}`}
      title={name}
      width="24"
      height="24"
      className="entity-icon"
      src={iconSrc}
    />
  );
};

export const generateRelationIconImage = (
  applicationName: string,
  applications: ModelData["applications"]
) => {
  const application = applications[applicationName];
  if (!application || !applicationName) {
    return;
  }
  return generateIconImg(applicationName, application["charm-url"]);
};

export const formatFriendlyDateToNow = (date: string) => {
  const formattedDate = formatDistanceToNow(parseISO(date));
  return formattedDate.concat(" ago");
};
