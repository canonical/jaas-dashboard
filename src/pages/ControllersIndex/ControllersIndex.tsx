import { Icon, MainTable, Tooltip } from "@canonical/react-components";
import {
  MainTableCell,
  MainTableHeader,
} from "@canonical/react-components/dist/components/MainTable/MainTable";
import FadeIn from "animations/FadeIn";
import cloneDeep from "clone-deep";
import Header from "components/Header/Header";
import { useQueryParams } from "hooks/useQueryParams";
import useWindowTitle from "hooks/useWindowTitle";
import BaseLayout from "layout/BaseLayout/BaseLayout";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { getControllerData, getModelData } from "store/juju/selectors";
import { AdditionalController, Controller } from "store/juju/types";
import ControllersOverview from "./ControllerOverview/ControllerOverview";
import "./_controllers.scss";

type AnnotatedController = (Controller | AdditionalController) & {
  models: number;
  machines: number;
  applications: number;
  units: number;
  wsControllerURL: string;
};

function Details() {
  useWindowTitle("Controllers");
  const controllerData = useSelector(getControllerData);
  const modelData = useSelector(getModelData);

  const controllerMap: Record<string, AnnotatedController> = {};
  const additionalControllers: string[] = [];
  if (controllerData) {
    Object.entries(controllerData).forEach(
      ([wsControllerURL, controllers], i) => {
        controllers.forEach((controller) => {
          const id =
            "uuid" in controller ? controller.uuid : `${wsControllerURL}-${i}`;
          if (controller.additionalController) {
            additionalControllers.push(id);
          }
          controllerMap[id] = {
            ...controller,
            models: 0,
            machines: 0,
            applications: 0,
            units: 0,
            wsControllerURL,
          };
        });
      }
    );
    if (modelData) {
      for (const modelUUID in modelData) {
        const model = modelData[modelUUID];
        if (model.info) {
          const controllerUUID = model?.info["controller-uuid"];
          if (controllerMap[controllerUUID]) {
            controllerMap[controllerUUID].models += 1;
            controllerMap[controllerUUID].machines += Object.keys(
              model?.machines
            ).length;
            const applicationKeys = Object.keys(model.applications);
            controllerMap[controllerUUID].applications +=
              applicationKeys.length;
            const unitCount = applicationKeys.reduce((acc, appName) => {
              const units = model.applications[appName].units || {}; // Subordinates don't have units
              return acc + Object.keys(units).length;
            }, 0);
            controllerMap[controllerUUID].units += unitCount;
          }
        }
      }
    }
  }

  const headers: MainTableHeader[] = [
    { content: "default", sortKey: "name" },
    { content: "cloud/region", sortKey: "cloud/region" },
    { content: "models", sortKey: "models" },
    { content: "machines", sortKey: "machines" },
    {
      content: "applications",
      sortKey: "applications",
    },
    { content: "units", sortKey: "units" },
    { content: "version", sortKey: "version" },
    { content: "access", sortKey: "public" },
  ];

  const additionalHeaders = cloneDeep(headers);
  additionalHeaders[0].content = (
    <span>
      Registered
      <span
        className="controllers--registered-tooltip p-icon--help"
        title="The controller authentication data is only stored in your browser localStorage. If you'd like this to persist across browsers try JAAS"
      ></span>
    </span>
  );

  function generatePathValue(controllerData: AnnotatedController) {
    const column: MainTableCell = { content: "" };
    if ("path" in controllerData && controllerData?.path === "admin/jaas") {
      column.content = "JAAS";
    } else if ("path" in controllerData && controllerData.path) {
      column.content = controllerData.path;
    } else {
      column.content = controllerData?.wsControllerURL;
      column.className = "is-disconnected";
      column.title = "disconnected";
    }
    return column;
  }

  function generateRow(c: AnnotatedController) {
    const cloud =
      "location" in c && c?.location?.cloud ? c.location.cloud : "unknown";
    const region =
      "location" in c && c?.location?.region ? c.location.region : "unknown";
    const cloudRegion = `${cloud}/${region}`;
    const access = "Public" in c && c.Public ? "Public" : "Private";
    let columns = [
      generatePathValue(c),
      { content: cloudRegion },
      { content: c.models },
      { content: c.machines },
      { content: c.applications },
      { content: c.units },
      { content: "" },
      { content: access, className: "u-capitalise" },
    ];
    if ("version" in c && c.version) {
      columns[columns.length - 2] = {
        content: (
          <>
            {c.version}{" "}
            {c.updateAvailable ? (
              <Tooltip
                message={
                  <>
                    There is an update or migration available for this
                    controller.{" "}
                    <a
                      className="p-list__link"
                      href="https://juju.is/docs/olm/upgrading"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      Read more
                    </a>
                  </>
                }
              >
                <Icon name="warning" data-testid="update-available" />
              </Tooltip>
            ) : null}
          </>
        ),
      };
    }
    return {
      columns,
    };
  }
  // XXX this isn't a great way of doing this.
  const additionalRows = additionalControllers.map((uuid) => {
    const row = generateRow(controllerMap[uuid]);
    delete controllerMap[uuid];
    return row;
  });

  const rows = controllerMap && Object.values(controllerMap).map(generateRow);

  const [, setPanelQs] = useQueryParams<{ panel: string | null }>({
    panel: null,
  });

  return (
    <>
      <div className="controllers--header">
        <div className="controllers__heading">
          Model status across controllers
        </div>
        <div className="controllers--register">
          <button
            className="p-button--positive"
            onClick={() => setPanelQs({ panel: "register-controller" })}
          >
            Register a controller
          </button>
        </div>
      </div>
      <ControllersOverview />
      <div className="l-controllers-table u-overflow--auto">
        {rows.length > 0 && <MainTable headers={headers} rows={rows} />}
        {additionalRows.length > 0 && (
          <MainTable headers={additionalHeaders} rows={additionalRows} />
        )}
      </div>
    </>
  );
}

export default function ControllersIndex() {
  const controllerData = useSelector(getControllerData);
  let controllerCount = 0;
  if (controllerData) {
    controllerCount = Object.keys(controllerData).length;
  }
  const modelData = useSelector(getModelData);
  let modelCount = 0;
  if (modelData) {
    modelCount = Object.keys(modelData).length;
  }

  return (
    <BaseLayout>
      <Header>
        <div className="entity-details__header">
          <strong className="controllers--count">
            {controllerCount} controllers,{" "}
            <Link to="/models">{modelCount} models</Link>
          </strong>
        </div>
      </Header>
      <div className="l-content controllers">
        <FadeIn isActive={true}>
          <Details />
        </FadeIn>
      </div>
    </BaseLayout>
  );
}
