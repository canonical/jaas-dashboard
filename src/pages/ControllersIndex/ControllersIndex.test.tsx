import { screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import type { RootState } from "store/store";
import { generalStateFactory } from "testing/factories/general";
import {
  additionalControllerFactory,
  controllerFactory,
  jujuStateFactory,
  modelDataApplicationFactory,
  modelDataFactory,
  modelDataInfoFactory,
  modelDataUnitFactory,
  modelDataMachineFactory,
  controllerInfoFactory,
} from "testing/factories/juju/juju";
import { connectionInfoFactory } from "testing/factories/juju/jujulib";
import { rootStateFactory } from "testing/factories/root";
import { renderComponent } from "testing/utils";

import ControllersIndex, { Label } from "./ControllersIndex";

describe("Controllers table", () => {
  let state: RootState;

  beforeEach(() => {
    state = rootStateFactory.withGeneralConfig().build();
  });

  it("renders a blank page if no data", () => {
    renderComponent(<ControllersIndex />, { state });
    expect(screen.queryByRole("row")).not.toBeInTheDocument();
  });

  it("renders the correct number of rows", () => {
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build(),
          controllerFactory.build({ path: "admin/jaas2", uuid: "456" }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(screen.getAllByRole("row")).toHaveLength(3);
  });

  it("displays additional controllers", () => {
    state.general.controllerConnections = {
      "wss://jimm.jujucharms.com/api": connectionInfoFactory.build({}),
    };
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build(),
          additionalControllerFactory.build(),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    const tables = screen.getAllByRole("grid");
    expect(tables).toHaveLength(2);
    const additionalRows = within(tables[1]).getAllByRole("row");
    expect(additionalRows).toHaveLength(2);
    expect(additionalRows[1]).toHaveTextContent(
      [
        "jimm.jujucharms.com",
        "Connected",
        "unknown/unknown",
        "0",
        "0",
        "0",
        "0",
      ].join(""),
    );
  });

  it("displays additional controllers from JIMM", () => {
    state.general.controllerConnections = {
      "wss://jimm.jujucharms.com/api": connectionInfoFactory.build(),
    };
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build(),
          controllerInfoFactory.build({ additionalController: true }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    const tables = screen.getAllByRole("grid");
    expect(tables).toHaveLength(2);
    const additionalRows = within(tables[1]).getAllByRole("row");
    expect(additionalRows).toHaveLength(2);
    expect(additionalRows[1]).toHaveTextContent(
      [
        "controller1",
        "Connected",
        "unknown/unknown",
        "0",
        "0",
        "0",
        "0",
        "1.2.3",
      ].join(""),
    );
  });

  it("counts models, machines, apps, and units", () => {
    state.general.controllerConnections = {
      "wss://jimm.jujucharms.com/api": connectionInfoFactory.build(),
    };
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [controllerFactory.build()],
      },
      modelData: {
        abc123: modelDataFactory.build({
          info: modelDataInfoFactory.build(),
          machines: {
            "0": modelDataMachineFactory.build(),
            "1": modelDataMachineFactory.build(),
          },
          applications: {
            easyrsa: modelDataApplicationFactory.build({
              units: {
                "easyrsa/0": modelDataUnitFactory.build(),
                "easyrsa/1": modelDataUnitFactory.build(),
              },
            }),
          },
        }),
        def456: modelDataFactory.build({
          info: modelDataInfoFactory.build(),
          machines: {
            "0": modelDataMachineFactory.build(),
            "1": modelDataMachineFactory.build(),
            "2": modelDataMachineFactory.build(),
          },
          applications: {
            ceph: modelDataApplicationFactory.build({
              units: {
                "ceph/0": modelDataUnitFactory.build(),
              },
            }),
          },
        }),
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(screen.getAllByRole("row")[1]).toHaveTextContent(
      [
        "admin/jaas",
        "Connected",
        "unknown/unknown",
        "2",
        "5",
        "2",
        "3",
        "1.2.3",
      ].join(""),
    );
  });

  it("displays login errors", async () => {
    state.general.loginErrors = {
      "wss://jimm.jujucharms.com/api": "Uh oh!",
    };

    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [controllerFactory.build()],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(
      screen.getByRole("gridcell", { name: "Failed to connect" }),
    ).toBeInTheDocument();
    await userEvent.hover(screen.getByText("Failed to connect"));
    expect(screen.getByText("Uh oh!")).toBeInTheDocument();
  });

  it("displays authentication required", async () => {
    state.general.controllerConnections = {};
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [controllerFactory.build()],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(
      screen.getByRole("gridcell", { name: "Authentication required" }),
    ).toBeInTheDocument();
  });

  it("shows 'Register new controller' panel", async () => {
    renderComponent(<ControllersIndex />, { url: "/controllers" });
    await userEvent.click(
      screen.getByRole("button", { name: Label.REGISTER_BUTTON }),
    );
    expect(window.location.search).toBe("?panel=register-controller");
    expect(
      document.querySelector(".p-panel.register-controller"),
    ).toBeInTheDocument();
  });

  it("Indicates if a controller has an update available", () => {
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build({
            uuid: "123",
            version: "1.0.0",
            updateAvailable: true,
          }),
          controllerFactory.build({
            uuid: "234",
            version: "1.0.0",
            updateAvailable: true,
          }),
          controllerFactory.build({
            uuid: "345",
            version: "0.9.9",
            updateAvailable: false,
          }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(screen.getAllByTestId("update-available")).toHaveLength(2);
  });

  it("displays notifications if controllers need authentication", () => {
    state.general = generalStateFactory.withConfig().build({
      visitURLs: ["/auth"],
    });
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build({
            uuid: "123",
            version: "1.0.0",
            updateAvailable: true,
          }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    expect(
      screen.getByText(/Controller authentication required/),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("link", { name: "Authenticate" }),
    ).toBeInTheDocument();
  });

  it("displays correct cloud data for Juju controller", () => {
    state.general.controllerConnections = {
      "wss://jimm.jujucharms.com/api": connectionInfoFactory.build(),
    };
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerFactory.build({
            location: { cloud: "cloud-aws", region: "eu-central-1" },
          }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    const tables = screen.getAllByRole("grid");
    const controllerRows = within(tables[0]).getAllByRole("row");
    expect(controllerRows).toHaveLength(2);
    expect(controllerRows[1]).toHaveTextContent(
      [
        "admin/jaas",
        "Connected",
        "cloud-aws/eu-central-1",
        "0",
        "0",
        "0",
        "0",
        "1.2.3",
      ].join(""),
    );
  });

  it("displays correct cloud data for JIMM enabled controller", () => {
    state.general.controllerConnections = {
      "wss://jimm.jujucharms.com/api": connectionInfoFactory.build(),
    };
    state.juju = jujuStateFactory.build({
      controllers: {
        "wss://jimm.jujucharms.com/api": [
          controllerInfoFactory.build({
            "cloud-tag": "cloud-aws",
            "cloud-region": "eu-central-1",
          }),
        ],
      },
    });
    renderComponent(<ControllersIndex />, { state });
    const tables = screen.getAllByRole("grid");
    const controllerRows = within(tables[0]).getAllByRole("row");
    expect(controllerRows).toHaveLength(2);
    expect(controllerRows[1]).toHaveTextContent(
      [
        "controller1",
        "Connected",
        "cloud-aws/eu-central-1",
        "0",
        "0",
        "0",
        "0",
        "1.2.3",
      ].join(""),
    );
  });
});
