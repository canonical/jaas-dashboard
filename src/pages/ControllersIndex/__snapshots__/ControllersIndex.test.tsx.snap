// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controllers table counts models, machines, apps, and units 1`] = `
<tr
  role="row"
>
  <td
    aria-hidden="false"
    class=""
    role="gridcell"
  >
    JAAS
  </td>
  <td
    aria-hidden="false"
    class=""
    role="gridcell"
  >
    unknown/unknown
  </td>
  <td
    aria-hidden="false"
    class="u-align--right"
    role="gridcell"
  >
    0
  </td>
  <td
    aria-hidden="false"
    class="u-align--right"
    role="gridcell"
  >
    0
  </td>
  <td
    aria-hidden="false"
    class="u-align--right"
    role="gridcell"
  >
    0
  </td>
  <td
    aria-hidden="false"
    class="u-align--right"
    role="gridcell"
  >
    0
  </td>
  <td
    aria-hidden="false"
    class="u-align--right"
    role="gridcell"
  />
  <td
    aria-hidden="false"
    class="u-align--right u-capitalise"
    role="gridcell"
  >
    False
  </td>
</tr>
`;
